server/aesdsocket.c

```diff
@@ -286,4 +290,170 @@ void create_daemon(void) {
+		syslog(LOG_INFO, "servinfo freed \r\n");
+	}
+
+	if (run_as_daemon == true)
+	{
+		create_daemon();
+	}
+
+	setup_signal_handlers();
+
+	if (listen(sock_fd, BACKLOG) == -1) {
+		syslog(LOG_ERR, "listen() failed with an error: %s\r\n", strerror(errno));
+		exit(EXIT_FAILURE);
+	}
+
+	// Initialize the link list
+    SLIST_HEAD(slist_head, thread_data_s) head;
+    SLIST_INIT(&head);
+
+	file_fd = open(FILE_NAME, O_CREAT | O_RDWR | O_APPEND, S_IWUSR | S_IRUSR | S_IRGRP | S_IROTH);
+	if(file_fd == -1) {
+		syslog(LOG_ERR, "Failed opening file %s with an error: %s", FILE_NAME, strerror(errno));
+		exit(EXIT_FAILURE);
+	}
+	syslog(LOG_INFO, "File %s opened for dumping the packet data\r\n", FILE_NAME);
+
+	while(!term_sig_received) {  // main accept() loop
+		struct sockaddr_in client_addr;
+		socklen_t client_addr_len = sizeof(client_addr);
+		accept_fd = accept(sock_fd, (struct sockaddr *)&client_addr, &client_addr_len);
+		if (accept_fd == -1) {
+			syslog(LOG_ERR, "accept() failed with an error: %s\r\n", strerror(errno));
+			continue;
+		}
+
+		char client_ip[INET_ADDRSTRLEN];
+		inet_ntop(PF_INET, &(client_addr.sin_addr), client_ip, INET_ADDRSTRLEN );
+		syslog(LOG_INFO, "Accepted connection from %s\r\n", client_ip);
+
+		struct thread_data_s *thread_struct = (struct thread_data_s *) malloc(sizeof(struct thread_data_s));
+        if (!thread_struct)
+        {
+            syslog(LOG_ERR, "Failed to allocate memory for thread structure\r\n");
+            exit(EXIT_FAILURE);
+        }
+        thread_struct->client_ip = client_ip;
+        thread_struct->client_fd = accept_fd;
+        thread_struct->thread_completed = false;        
+        int thread_id = pthread_create(&thread, NULL, thread_func, thread_struct);        
+        if (thread_id != 0)
+        {
+            syslog(LOG_ERR, "Error creating new thread");
+            free(thread_struct);
+            exit(EXIT_FAILURE);
+        }
+        thread_struct->thread_id = thread;
+        SLIST_INSERT_HEAD(&head, thread_struct, entries);
+
+        struct thread_data_s *thread_ptr = NULL;
+        struct thread_data_s *next_thread = NULL;
+        SLIST_FOREACH_SAFE(thread_ptr, &head, entries, next_thread)
+        {
+	        if (thread_ptr->thread_completed)
+	        {
+	            syslog(LOG_INFO, "Thread complete, joining");
+	            int id = pthread_join(thread_ptr->thread_id, NULL);
+	            if (id != 0)
+	            {
+	                syslog(LOG_ERR, "Failure joining thread");
+	                exit(EXIT_FAILURE);
+	            }
+	            SLIST_REMOVE(&head, thread_ptr, thread_data_s, entries);
+	            free(thread_ptr);
+	        }
+	    }
+	}
+
+    while (!SLIST_EMPTY(&head))
+    {
+        struct thread_data_s *thread_rm = SLIST_FIRST(&head);
+        pthread_join(thread_rm->thread_id, NULL);
+        SLIST_REMOVE_HEAD(&head, entries);
+        free(thread_rm);
+    }
+
+    pthread_mutex_destroy(&mutex);
+	//free(recv_buf);
+	close(accept_fd);
```

Comment:
@krishnasuhagiya29 valgrind output [here](https://github.com/cu-ecen-aeld/assignments-3-and-later-krishnasuhagiya29/pull/4/files#diff-85261e472d268426712db558b85101dca2eb5ab21de43bcc50c1450ee339e7ecR114) is pointing out that you are calling close with a -1 file descriptor after accept() fails due to signal and shutdown.